lib/ipam/ipam.go

```diff
@@ -548,7 +549,44 @@ func (s *blockAssignState) findOrClaimBlock(ctx context.Context, minFreeIps int)
 	return nil, false, errors.New("failed to find or claim a block")
 }
 
-func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, requestedPools []net.IPNet, version int, host string, maxNumBlocks int, rsvdAttr *HostReservedAttr) ([]net.IPNet, error) {
+type IPAMAssignments struct {
+	IPs              []net.IPNet       // assigned IP addresses
+	IPVersion        int               // IP version (4 or 6)
+	NumRequested     int               // number of requested IP addresses (not all may be assigned)
+	HostReservedAttr *HostReservedAttr // reserved addresses at start and/or end of blocks
+	Msgs             []string          // warning/error messages to be rendered in case there are any issues with the assignment
+}
+
+func (i *IPAMAssignments) AddMsg(msg string) {
+	for _, m := range i.Msgs {
```

Comment:
nit: might be more efficient to put this in a map so we don't need to do a loop through all of them to check. It might not matter too much either way since I don't expect this list to really get super long usually.

Reply:
I'm also wondering if we're only using `Msgs` to just keep a list of additional info if we should even store it as separate strings or if we should just keep it as a string builder and then just `WriteString` the new messages to it. As far as I can tell we only want to keep them separate so we can do the comparison to see if we've already seen the error message before right?

Reply:
@caseydavenport 's reasoning for keeping the string list is to have the "breadcrumb" information of what error/warning was hit until we get into the current state. So a plain map would lose that ordering, I had considered a map with some kind of helper structure to keep the ordering, but I didn't think we would have too great a number of msgs that would make the `for` loop too bad on performance... Yes, a string builder should work, but I'm not sure about filtering duplicates with it, would there be an efficient way of doing this?